File: 75.sort-colors.java
/*
 * @lc app=leetcode id=75 lang=java
 *
 * [75] Sort Colors
 */
class Solution {
    public void sortColors(int[] nums) {
        if (nums == null) return;
        int zeroPtr = 0, twoPtr = nums.length-1;
        
        for (int onePtr = 0; onePtr <= twoPtr && zeroPtr < twoPtr;) {
            if (nums[onePtr] == 0) {
                nums[onePtr] = nums[zeroPtr];
                nums[zeroPtr++] = 0;
                if (onePtr < zeroPtr) onePtr = zeroPtr;
            } else if (nums[onePtr] == 2) {
                nums[onePtr] = nums[twoPtr];
                nums[twoPtr--] = 2;
            } else {
                onePtr++;
            }
        }
    }
}

File: go/75.sort-colors.go
/*
 * @lc app=leetcode id=75 lang=golang
 *
 * [75] Sort Colors
 */

// @lc code=start
func sortColors(nums []int) {
	l, r := 0, len(nums)-1
	for i := 0; i <= r && l < r; {
		switch nums[i] {
		case 0:
			nums[i] = nums[l]
			nums[l] = 0
			l++
			if i < l {
				i = l
			}
		case 2:
			nums[i] = nums[r]
			nums[r] = 2
			r--
		default:
			i++
		}
	}
}

// @lc code=end

